src/work/BasicWork.cpp

```diff
@@ -0,0 +1,262 @@
+                     std::string name, size_t maxRetries)
+    : mApp(app)
+    , mNotifyCallback(std::move(callback))
+    , mName(std::move(name))
+    , mMaxRetries(maxRetries)
+{
+}
+
+BasicWork::~BasicWork()
+{
+    mState = WORK_DESTRUCTING;
+}
+
+std::string const&
+BasicWork::getName() const
+{
+    return mName;
+}
+
+std::string
+BasicWork::getStatus() const
+{
+    // Work is in `WAITING` state when retrying
+    auto state = mRetryTimer ? WORK_FAILURE_RETRY : mState;
+
+    switch (state)
+    {
+    case WORK_RUNNING:
+        return fmt::format("Running: {:s}", getName());
+    case WORK_WAITING:
+        return fmt::format("Waiting: {:s}", getName());
+    case WORK_SUCCESS:
+        return fmt::format("Succeeded: {:s}", getName());
+    case WORK_FAILURE_RETRY:
+    {
+        auto eta = getRetryETA();
+        return fmt::format("Retrying in {:d} sec: {:s}", eta, getName());
+    }
+    case WORK_FAILURE_RAISE:
+        return fmt::format("Failed: {:s}", getName());
+    default:
+        assert(false);
+        return "";
+    }
+}
+
+bool
+BasicWork::isDone() const
+{
+    return mState == WORK_SUCCESS || mState == WORK_FAILURE_RAISE;
+}
+
+std::string
+BasicWork::stateName(State st)
+{
+    switch (st)
+    {
+    case WORK_WAITING:
+        return "WORK_WAITING";
+    case WORK_RUNNING:
+        return "WORK_RUNNING";
+    case WORK_SUCCESS:
+        return "WORK_SUCCESS";
+    case WORK_FAILURE_RETRY:
+        return "WORK_FAILURE_RETRY";
+    case WORK_FAILURE_RAISE:
+        return "WORK_FAILURE_RAISE";
+    default:
+        throw std::runtime_error("Unknown Work::State");
+    }
+}
+
+void
+BasicWork::reset()
+{
+    CLOG(DEBUG, "Work") << "resetting " << getName();
+    setState(WORK_RUNNING);
+    onReset();
+}
+
+std::function<void()>
+BasicWork::wakeUpCallback()
+{
+    std::weak_ptr<BasicWork> weak = shared_from_this();
+    auto callback = [weak]() {
+        auto self = weak.lock();
+        if (self)
+        {
+            self->wakeUp();
+        }
+    };
+    return callback;
+}
+
+void
+BasicWork::waitForRetry()
+{
+    if (mRetryTimer)
+    {
+        throw std::runtime_error(
+            fmt::format("Retry timer for {} already exists!", getName()));
+    }
+
+    mRetryTimer = std::make_unique<VirtualTimer>(mApp.getClock());
+    std::weak_ptr<BasicWork> weak = shared_from_this();
+    auto t = getRetryDelay();
+    mRetryTimer->expires_from_now(t);
+    CLOG(WARNING, "Work")
+        << "Scheduling retry #" << (mRetries + 1) << "/" << mMaxRetries
+        << " in " << std::chrono::duration_cast<std::chrono::seconds>(t).count()
+        << " sec, for " << getName();
+    setState(WORK_WAITING);
+    mRetryTimer->async_wait(
+        [weak]() {
+            auto self = weak.lock();
+            if (!self)
+            {
+                return;
+            }
+            self->mRetries++;
+            self->mRetryTimer = nullptr;
+            self->wakeUp();
```

Comment:
this is wrong:
* first you should do whatever you need to do, which, in this case is "transition to RUNNING", and then wakeup.
* wakeup should be called by `setState` as the expectation is that if you get out of `WAITING` state, you call it